### PR TITLE
Add examples to illustrate use of fill_include/exclude_names

### LINKED ISSUE #10001: Add example of ascii fill_include/exclude_names
### Description

To provide better illumination on how `fill_include_names` and `fill_exclude_names` work, add a doc example similar to below:

```
>>> lines = ['a,b,c,d', '1,2,3,4', ',,,']

>>> ascii.read(lines)
<Table length=2>
  a     b     c     d  
int64 int64 int64 int64
----- ----- ----- -----
    1     2     3     4
   --    --    --    --
>>> ascii.read(lines, fill_include_names=['a', 'c'])
<Table length=2>
  a    b     c    d  
int64 str1 int64 str1
----- ---- ----- ----
    1    2     3    4
   --         --     

>>> ascii.read(lines, fill_exclude_names=['a', 'c'])
<Table length=2>
 a     b    c     d  
str1 int64 str1 int64
---- ----- ---- -----
   1     2    3     4
        --         --
```

### PR DESCRIPTION
**Pull request by kakirastern**
### Description

This pull request is to address the lack of proper documentation for the usage of the `fill_include_names` and `fill_exclude_names` attributes of the `BaseData` class of `astropy.io.ascii` for the input/output of ASCII files. Some newly written material, both describing their usage and give some basic examples, has been added to the `docs/io/ascii/index.rst` file in the `astropy` (this) repo for ease of reference. 

Fixes #10001.

### CODE REVIEW

File: docs/io/ascii/index.rst [hunk 1]
```diff
@@ -229,6 +229,63 @@ ASCII with no loss of information.
...
+but for a string column one might desire those to empty strings "" to be not masked.
+Or perhaps one has a file that uses the number -99 as the missing data indicator
+for most columns but then -99 occurs as real data in another column. These features
+allow the data to be read in with ``astropy.io.ascii`` correctly without too much effort.
+
+Examples
+^^^^^^^^
+..
+  EXAMPLE START
+  Using the ``fill_include_names`` and ``fill_exclude_names`` attributes for ASCII tables
```
Comment (bsipocz) [inline docs/io/ascii/index.rst]: these are parameters rather than attributes, please update the usage of the word in this patch.

Comment (kakirastern) [reply docs/io/ascii/index.rst]: Sure, I have updated the wording from `attribute(s)` to `parameter(s)` in the wording.

**Comment by kakirastern** [comment]
I experimented with using single backticks as a check. Will revert back to using double backticks for the parameters.

File: docs/io/ascii/index.rst [hunk 2]
```diff
@@ -229,6 +229,63 @@ ASCII with no loss of information.
...
+Some points of interests pertaining to masking are the ``fill_include_names``
+and ``fill_exclude_names`` parameters of the :class:`~astropy.io.ascii.BaseData` class.
+The original intention for the usage of these parameters are in the ``masks``
+method of the same :class:`~astropy.io.ascii.BaseData` class. In the first step of the
+``masks`` method, the mask is evaluated with values from ``fill_values`` as applied
+to which column using ``fill_include_names`` and ``fill_exclude_names``.
+In the second step, all the replacements of values are done for the appropriate columns.
+As standalone tools, they can be used to serve in isolating columns that are of interests
+while not deleting them outright. We can use the ``fill_include_names`` parameter in include
+only columns with names as indicated. On the contrary but highly similarly, we can use
```
Comment (lglattly) [inline docs/io/ascii/index.rst]:
```suggestion
only columns with names as indicated. Similarly, we can use
```

File: docs/io/ascii/index.rst [hunk 3]
```diff
@@ -229,6 +229,63 @@ ASCII with no loss of information.
...
+the ``fill_exclude_names`` parameter to exclude the columns with the names as indicated.
+One use case is where one has a CSV file with empty values indicating missing values,
```
Comment (lglattly) [inline docs/io/ascii/index.rst]:
```suggestion
One use case is where a user has a CSV file with empty values indicating missing values,
```

File: docs/io/ascii/index.rst [hunk 4]
```diff
@@ -229,6 +229,63 @@ ASCII with no loss of information.
...
+but for a string column one might desire those to empty strings "" to be not masked.
```
Comment (lglattly) [inline docs/io/ascii/index.rst]:
```suggestion
but for a string column they might want those empty strings "" to not be masked.
```

File: docs/io/ascii/index.rst [hunk 5]
```diff
@@ -229,6 +229,63 @@ ASCII with no loss of information.
...
+Or perhaps one has a file that uses the number -99 as the missing data indicator
```
Comment (lglattly) [inline docs/io/ascii/index.rst]:
```suggestion
Or perhaps you have a file that uses the number -99 as the missing data indicator
```

File: docs/io/ascii/index.rst [hunk 6]
```diff
@@ -229,6 +229,63 @@ ASCII with no loss of information.
...
+for most columns but then -99 occurs as real data in another column. These features
```
Comment (lglattly) [inline docs/io/ascii/index.rst]:
```suggestion
for most columns, but then -99 occurs as real data in another column. These features
```

File: docs/io/ascii/index.rst [hunk 7]
```diff
@@ -229,6 +229,63 @@ ASCII with no loss of information.
...
+allow the data to be read in with ``astropy.io.ascii`` correctly without too much effort.
+
+Examples
+^^^^^^^^
+..
+  EXAMPLE START
+  Using the ``fill_include_names`` and ``fill_exclude_names`` parameters for ASCII tables
```
Comment (lglattly) [inline docs/io/ascii/index.rst]:
```suggestion
  Using the ``fill_include_names`` and ``fill_exclude_names`` Parameters for ASCII Tables
```

File: docs/io/ascii/index.rst [hunk 8]
```diff
@@ -229,6 +229,63 @@ ASCII with no loss of information.
...
+
+The following gives a simple example to illustrate how ``fill_include_names``
```
Comment (lglattly) [inline docs/io/ascii/index.rst]:
```suggestion
The following illustrates how ``fill_include_names``
```

**Comment by bsipocz** [comment]
> I experimented with using single backticks as a check. Will revert back to using double backticks for the parameters.

single backticks are always tried to be resolved by sphinx to references, API links, etc. there are no anchors on parameters to link to, so there is really no need to experiment with it using the main astropy CI. Instead t I encourage you to look into things like this locally if you like.
 (If you would like to learn more about docs generation, a next step in getting a good understanding of the logic is to look into the intersphinx mapping files directly, as that is the one helping sphinx to resolve the links. Looking into the sphinx code of course also shed some light on the logic, but that's very much an advanced and not at all intuitive way to look at things, and a source of infinite rabbit holes and timesinks...).

File: docs/io/ascii/index.rst [hunk 9]
```diff
@@ -229,6 +229,63 @@ ASCII with no loss of information.
 ..
   EXAMPLE END
 
+Miscellaneous Examples
```
Comment (taldcroft) [inline docs/io/ascii/index.rst]: This should be a subsection of https://astropy.readthedocs.io/en/latest/io/ascii/read.html#bad-or-missing-values and should be titled `Selecting columns for masking`.

File: docs/io/ascii/index.rst [hunk 10]
```diff
@@ -229,6 +229,63 @@ ASCII with no loss of information.
...
+and ``fill_exclude_names`` parameters of the :class:`~astropy.io.ascii.BaseData` class.
+The original intention for the usage of these parameters are in the ``masks``
+method of the same :class:`~astropy.io.ascii.BaseData` class. In the first step of the
+``masks`` method, the mask is evaluated with values from ``fill_values`` as applied
+to which column using ``fill_include_names`` and ``fill_exclude_names``.
+In the second step, all the replacements of values are done for the appropriate columns.
+As standalone tools, they can be used to serve in isolating columns that are of interests
+while not deleting them outright. We can use the ``fill_include_names`` parameter in include
+only columns with names as indicated. On the contrary but highly similarly, we can use
+the ``fill_exclude_names`` parameter to exclude the columns with the names as indicated.
```
Comment (taldcroft) [inline docs/io/ascii/index.rst]: This whole section is a bit too detailed and hard to follow.  I'd suggest just boiling it down to this:
```
The |read| function provides parameters ``fill_include_names`` and ``fill_exclude_names``
to select which columns will be used in the ``fill_values`` masking process described above.
The use of these parameters is not common but in some cases can considerably simplify
the code required to read a table.
```

Comment (kakirastern) [reply docs/io/ascii/index.rst]: Sure! I have made some changes according to the suggestions in the most recent commit(s).

File: docs/io/ascii/index.rst (hunk 9 above)
Comment (kakirastern) [reply docs/io/ascii/index.rst]: Sure, have moved the changes to the [io/ascii/read.html#bad-or-missing-values](https://astropy.readthedocs.io/en/latest/io/ascii/read.html#bad-or-missing-values) section instead.

File: docs/io/ascii/index.rst (hunk 8 above)
Comment (kakirastern) [reply docs/io/ascii/index.rst]: Thanks @lglattly for the suggested edits!

**Comment by kakirastern** [comment]
> single backticks are always tried to be resolved by sphinx to references, API links, etc. there are no anchors on parameters to link to, so there is really no need to experiment with it using the main astropy CI. Instead t I encourage you to look into things like this locally if you like.
> (If you would like to learn more about docs generation, a next step in getting a good understanding of the logic is to look into the intersphinx mapping files directly, as that is the one helping sphinx to resolve the links. Looking into the sphinx code of course also shed some light on the logic, but that's very much an advanced and not at all intuitive way to look at things, and a source of infinite rabbit holes and timesinks...).

Yeah, good suggestion! That would indeed be something I am looking forward to picking up on next. Thanks @bsipocz!

**Comment by kakirastern** [comment]
All CI tests passed. Ready for another review.

**Comment by kakirastern** [comment]
Fixed earlier omission by adding a subsection entitled `Selecting columns for masking` under the `Bad or Missing Values` section.

**Comment by kakirastern** [comment]
Hi, I am getting the following error while passing the `ci/circleci: 32bit-and-parallel` test by using `int32` instead of `int64`:

```
_______________________ [doctest] docs/io/ascii/read.rst _______________________
335   Using the ``fill_include_names`` and ``fill_exclude_names`` parameters for ASCII tables
336 
337 The use of these parameters is not common but in some cases can considerably simplify
338 the code required to read a table. The following gives a simple example to illustrate how
339 ``fill_include_names`` and ``fill_exclude_names`` can be used
340 in the most basic and typical cases::
341 
342   >>> from astropy.io import ascii
343   >>> lines = ['a,b,c,d', '1,2,3,4', ',,,']
344   >>> ascii.read(lines)
Differences (unified diff with -expected +actual):
    @@ -1,5 +1,5 @@
     <Table length=2>
    -  a     b     c     d
    -int32 int32 int32 int32
    +  a     b     c     d
    +int64 int64 int64 int64
     ----- ----- ----- -----
         1     2     3     4
```
Seems I tried the Travis CI suggested version before, but was failing the `ci/circleci: 32bit-and-parallel` checks. Seems like Travis CI and CircleCI are giving me contradictory checks. Is there any recommended course of action at this juncture? Should I simply skip the doc test by adding `# doctest: +SKIP` to this portion of the code (or code block)?

**Comment by pllim** [comment]
What about explicitly setting the `dtype`?

**Comment by kakirastern** [comment]
Just see your comments now @pllim. Let me try your approach as well.

**Comment by kakirastern** [comment]
I tried converting to dtype to `numpy.int` explicitly using the converter method, but keep getting `int64` locally. But this would not work since it would not necessarily pass the `32bit-and-parallel` CircleCI checks I reckon. It really depends, but I am not sure.

**Comment by pllim** [comment]
Huh, tests passed now. 🤷‍♀  Thanks!

**Comment by bsipocz** [comment]
> Huh, tests passed now. 🤷‍♀ Thanks!

passed as the doctest has been skipped, which is not what was suggested in the review 🤷‍♀

**Comment by kakirastern** [comment]
Sorry guys, did the tweaking based on a line of code in the same file before I saw @pllim's comment. I did try her `converter`approach locally, but it wasn't working on my machine.

**Comment by bsipocz** [comment]
what about using the dtype instead, as it's been done elsewhere in the table docs, where the dtypes are irrelevant for the sake of the things the actual example shows, like here.

**Comment by kakirastern** [comment]
> what about using the dtype instead, as it's been done elsewhere in the table docs, where the dtypes are irrelevant for the sake of the things the actual example shows, like here.
> 
> Skipping a doctest line should be the last resource to use when all other other hacks are failing.

Hi @bsipocz Could you share the link for the example again? The first one appears not to be showing...

**Comment by pllim** [comment]
@kakirastern , did this not work for you? @bsipocz , I cannot find `dtype` example for `io.ascii` (maybe that is just for the `table` interface?).

```python
>>> import numpy as np
>>> from astropy.io import ascii
>>> lines = ['a,b,c,d', '1,2,3,4', ',,,']
>>> converters = {'a': [ascii.convert_numpy(np.int32)], 'b': [ascii.convert_numpy(np.int32)], 'c': [ascii.convert_numpy(np.int32)], 'd': [ascii.convert_numpy(np.int32)]}
>>> ascii.read(lines, converters=converters)
```
```
<Table length=2>
  a     b     c     d  
int32 int32 int32 int32
----- ----- ----- -----
    1     2     3     4
   --    --    --    --
```

**Comment by bsipocz** [comment]
Yes, `dtype` is with the unified io interface, or column constructor. Otherwise I agree, converters should work as you suggested.

**Comment by kakirastern** [comment]
Tried something as @pllim has suggested, but now am getting the error as follows for Travis CI:  

```
=================================== FAILURES ===================================
_______________________ [doctest] docs/io/ascii/read.rst _______________________
345   >>> converters = {'a': [ascii.convert_numpy(np.int32)], 'b': [ascii.convert_numpy(np.int32)], 'c': [ascii.convert_numpy(np.int32)], 'd': [ascii.convert_numpy(np.int32)]}
346   >>> ascii.read(lines, converters=converters)
347   <Table length=2>
348     a     b     c     d
349   int32 int32 int32 int32
350   ----- ----- ----- -----
351       1     2     3     4
352      --    --    --    --
353 
354   >>> ascii.read(lines, fill_include_names=['a', 'c'], converters=converters)
Differences (unified diff with -expected +actual):
    @@ -1,6 +1,7 @@
    -<Table length=2>
    -  a    b     c    d
    -int32 str1 int32 str1
    ------ ---- ----- ----
    -    1    2     3    4
    -   --         --
    +<Table length=3>
    +col1 col2 col3 col4
    +str1 str1 str1 str1
    +---- ---- ---- ----
    +   a    b    c    d
    +   1    2    3    4
    +<BLANKLINE>

```

**Comment by pllim** [comment]
Thanks for trying it out, @kakirastern . Yeah, I don't know the subtleties enough for the second case to work with converters. This is what I get when I modified the converters a little. Any ideas, @taldcroft ?

```python
>>> import numpy as np
>>> from astropy.io import ascii
>>> lines = ['a,b,c,d', '1,2,3,4', ',,,']
>>> converters_col = {'col1': [ascii.convert_numpy(np.int32)], 'col2': [ascii.convert_numpy(np.int32)], 'col3': [ascii.convert_numpy(np.int32)], 'col4': [ascii.convert_numpy(np.int32)]}
>>> ascii.read(lines, fill_include_names=['a', 'c'], converters=converters_col)
```
```
<Table length=2>
  a    b     c    d  
int64 str1 int64 str1
----- ---- ----- ----
    1    2     3    4
   --         --     
```
I had to use the underlying "colN" naming for converters. Also, columns not listed in `fill_include_names` somehow is `str1`, not `int32` as indicated by converters. Is this a bug or a feature?

**Comment by kakirastern** [comment]
No worries @pllim If need be I could revert to doctest skipping. But that's only as a last resort.

**Comment by taldcroft** [comment]
@kakirastern - a fine workaround for the Windows annoyance is to use values like `1.0, 2.0, 3.0, 4.0` instead of `1, 2, 3, 4`.  Those will be float64 on all platforms and allow running the doc tests.

**Comment by kakirastern** [comment]
Hi @taldcroft Just made the changes from `int64` to `float64` as you have suggested.

**Comment by kakirastern** [comment]
My apologies about the `ascii` table formatting. Am using Jupyter notebook for the checking. Should try `ipython` next time.
